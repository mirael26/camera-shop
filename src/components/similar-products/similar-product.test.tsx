import { fireEvent, screen } from '@testing-library/react';
import axios from 'axios';
import { renderWithReduxAndRouter } from '../../test/helpers/render-with-redux-and-router';
import { bigProductsMock, productMock, productsMock } from '../../test/mocks';
import SimilarProducts from './similar-products';

jest.mock('axios');
jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({ id: '4' }),
  }
});

describe('SimilarProducts', () => {
  beforeEach(() => {
    jest.mocked(axios).get.mockResolvedValue(productsMock);
  });
  
  test('Render correctly', () => {
    const similarProducts = renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: productsMock }
      }
    });
    expect(similarProducts).toMatchSnapshot();
  });

  test('Make one get-query', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: productsMock }
      }
    });

    expect(jest.mocked(axios).get).toHaveBeenCalledTimes(1);
  });

  test('Render right cards count when 1 similar', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: [productMock] }
      }
    });

    const cards = screen.getAllByTestId('product-card');
    expect(cards).toHaveLength(1);
  });

  test('Render right cards count when 5 similar', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: bigProductsMock }
      }
    });

    const cards = screen.getAllByTestId('product-card');
    expect(cards).toHaveLength(5);
    expect(screen.getByLabelText('Предыдущий слайд')).toBeInTheDocument();
    expect(screen.getByLabelText('Следующий слайд')).toBeInTheDocument();
  });

  test('Render right visible cards count', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: bigProductsMock }
      }
    });

    const cards = screen.getAllByTestId('product-card');
    expect(cards).toHaveLength(5);
    expect(cards[0]).toHaveClass('is-active');
    expect(cards[1]).toHaveClass('is-active');
    expect(cards[2]).toHaveClass('is-active');
    expect(cards[3]).not.toHaveClass('is-active');
    expect(cards[4]).not.toHaveClass('is-active');
  });

  test('Dont render component when no similar products', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: [] }
      }
    });

    expect(screen.queryByTestId('similar-products')).not.toBeInTheDocument();
  });

  test('Dont render prev/next button when 1 card', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: [productMock] }
      }
    });

    expect(screen.queryByLabelText('Предыдущий слайд')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('Следующий слайд')).not.toBeInTheDocument();
  });

  test('Render right prev/next buttons state', () => {
    renderWithReduxAndRouter(<SimilarProducts/>, {
      initialState: {
        data: { similarProducts: bigProductsMock }
      }
    });

    const prevButton = screen.getByLabelText('Предыдущий слайд');
    const nextButton = screen.getByLabelText('Следующий слайд');

    expect(prevButton).toHaveAttribute('disabled');
    expect(nextButton).not.toHaveAttribute('disabled');

    fireEvent.click(nextButton);
    fireEvent.click(nextButton);
    
    expect(prevButton).not.toHaveAttribute('disabled');
    expect(nextButton).toHaveAttribute('disabled');
  });
});
