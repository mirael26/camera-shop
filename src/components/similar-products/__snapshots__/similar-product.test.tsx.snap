// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimilarProducts Render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page-content__section"
      >
        <section
          class="product-similar"
          data-testid="similar-products"
        >
          <div
            class="container"
          >
            <h2
              class="title title--h3"
            >
              Похожие товары
            </h2>
            <div
              class="product-similar__slider"
            >
              <div
                class="product-similar__slider-list"
              >
                <div
                  class="product-card is-active"
                  data-testid="product-card"
                >
                  <div
                    class="product-card__img"
                  >
                    <picture>
                      <source
                        srcset="img/content/orlenok.webp, img/content/orlenok@2x.webp 2x"
                        type="image/webp"
                      />
                      <img
                        alt="Орлёнок"
                        height="240"
                        src="img/content/orlenok.jpg"
                        srcset="img/content/orlenok@2x.jpg 2x"
                        width="280"
                      />
                    </picture>
                  </div>
                  <div
                    class="product-card__info"
                  >
                    <div
                      class="rate product-card__rate"
                    >
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          xlink:href="#icon-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          xlink:href="#icon-star"
                        />
                      </svg>
                      <p
                        class="visually-hidden"
                        data-testid="rating"
                      >
                        Рейтинг: 
                        3
                      </p>
                      <p
                        class="rate__count"
                        data-testid="rate-count"
                      >
                        <span
                          class="visually-hidden"
                        >
                          Всего оценок:
                        </span>
                        17
                      </p>
                    </div>
                    <p
                      class="product-card__title"
                      data-testid="name"
                    >
                      Орлёнок
                    </p>
                    <p
                      class="product-card__price"
                      data-testid="price"
                    >
                      <span
                        class="visually-hidden"
                      >
                        Цена:
                      </span>
                      19 970
                       ₽
                    </p>
                  </div>
                  <div
                    class="product-card__buttons"
                  >
                    <button
                      class="btn btn--purple product-card__btn"
                      type="button"
                    >
                      Купить
                    </button>
                    <a
                      class="btn btn--transparent"
                      href="/catalog/product/4"
                    >
                      Подробнее
                    </a>
                  </div>
                </div>
                <div
                  class="product-card is-active"
                  data-testid="product-card"
                >
                  <div
                    class="product-card__img"
                  >
                    <picture>
                      <source
                        srcset="img/content/van-shot.webp, img/content/van-shot@2x.webp 2x"
                        type="image/webp"
                      />
                      <img
                        alt="Van Shot"
                        height="240"
                        src="img/content/van-shot.jpg"
                        srcset="img/content/van-shot@2x.jpg 2x"
                        width="280"
                      />
                    </picture>
                  </div>
                  <div
                    class="product-card__info"
                  >
                    <div
                      class="rate product-card__rate"
                    >
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <p
                        class="visually-hidden"
                        data-testid="rating"
                      >
                        Рейтинг: 
                        5
                      </p>
                      <p
                        class="rate__count"
                        data-testid="rate-count"
                      >
                        <span
                          class="visually-hidden"
                        >
                          Всего оценок:
                        </span>
                        2
                      </p>
                    </div>
                    <p
                      class="product-card__title"
                      data-testid="name"
                    >
                      Van Shot
                    </p>
                    <p
                      class="product-card__price"
                      data-testid="price"
                    >
                      <span
                        class="visually-hidden"
                      >
                        Цена:
                      </span>
                      149 990
                       ₽
                    </p>
                  </div>
                  <div
                    class="product-card__buttons"
                  >
                    <button
                      class="btn btn--purple product-card__btn"
                      type="button"
                    >
                      Купить
                    </button>
                    <a
                      class="btn btn--transparent"
                      href="/catalog/product/5"
                    >
                      Подробнее
                    </a>
                  </div>
                </div>
                <div
                  class="product-card is-active"
                  data-testid="product-card"
                >
                  <div
                    class="product-card__img"
                  >
                    <picture>
                      <source
                        srcset="img/content/click-sap.webp, img/content/click-sap@2x.webp 2x"
                        type="image/webp"
                      />
                      <img
                        alt="Click Sap"
                        height="240"
                        src="img/content/click-sap.jpg"
                        srcset="img/content/click-sap@2x.jpg 2x"
                        width="280"
                      />
                    </picture>
                  </div>
                  <div
                    class="product-card__info"
                  >
                    <div
                      class="rate product-card__rate"
                    >
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          data-testid="full-star-icon"
                          xlink:href="#icon-full-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          xlink:href="#icon-star"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        height="16"
                        width="17"
                      >
                        <use
                          xlink:href="#icon-star"
                        />
                      </svg>
                      <p
                        class="visually-hidden"
                        data-testid="rating"
                      >
                        Рейтинг: 
                        3
                      </p>
                      <p
                        class="rate__count"
                        data-testid="rate-count"
                      >
                        <span
                          class="visually-hidden"
                        >
                          Всего оценок:
                        </span>
                        3
                      </p>
                    </div>
                    <p
                      class="product-card__title"
                      data-testid="name"
                    >
                      Click Sap
                    </p>
                    <p
                      class="product-card__price"
                      data-testid="price"
                    >
                      <span
                        class="visually-hidden"
                      >
                        Цена:
                      </span>
                      9 490
                       ₽
                    </p>
                  </div>
                  <div
                    class="product-card__buttons"
                  >
                    <button
                      class="btn btn--purple product-card__btn"
                      type="button"
                    >
                      Купить
                    </button>
                    <a
                      class="btn btn--transparent"
                      href="/catalog/product/6"
                    >
                      Подробнее
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page-content__section"
    >
      <section
        class="product-similar"
        data-testid="similar-products"
      >
        <div
          class="container"
        >
          <h2
            class="title title--h3"
          >
            Похожие товары
          </h2>
          <div
            class="product-similar__slider"
          >
            <div
              class="product-similar__slider-list"
            >
              <div
                class="product-card is-active"
                data-testid="product-card"
              >
                <div
                  class="product-card__img"
                >
                  <picture>
                    <source
                      srcset="img/content/orlenok.webp, img/content/orlenok@2x.webp 2x"
                      type="image/webp"
                    />
                    <img
                      alt="Орлёнок"
                      height="240"
                      src="img/content/orlenok.jpg"
                      srcset="img/content/orlenok@2x.jpg 2x"
                      width="280"
                    />
                  </picture>
                </div>
                <div
                  class="product-card__info"
                >
                  <div
                    class="rate product-card__rate"
                  >
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        xlink:href="#icon-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        xlink:href="#icon-star"
                      />
                    </svg>
                    <p
                      class="visually-hidden"
                      data-testid="rating"
                    >
                      Рейтинг: 
                      3
                    </p>
                    <p
                      class="rate__count"
                      data-testid="rate-count"
                    >
                      <span
                        class="visually-hidden"
                      >
                        Всего оценок:
                      </span>
                      17
                    </p>
                  </div>
                  <p
                    class="product-card__title"
                    data-testid="name"
                  >
                    Орлёнок
                  </p>
                  <p
                    class="product-card__price"
                    data-testid="price"
                  >
                    <span
                      class="visually-hidden"
                    >
                      Цена:
                    </span>
                    19 970
                     ₽
                  </p>
                </div>
                <div
                  class="product-card__buttons"
                >
                  <button
                    class="btn btn--purple product-card__btn"
                    type="button"
                  >
                    Купить
                  </button>
                  <a
                    class="btn btn--transparent"
                    href="/catalog/product/4"
                  >
                    Подробнее
                  </a>
                </div>
              </div>
              <div
                class="product-card is-active"
                data-testid="product-card"
              >
                <div
                  class="product-card__img"
                >
                  <picture>
                    <source
                      srcset="img/content/van-shot.webp, img/content/van-shot@2x.webp 2x"
                      type="image/webp"
                    />
                    <img
                      alt="Van Shot"
                      height="240"
                      src="img/content/van-shot.jpg"
                      srcset="img/content/van-shot@2x.jpg 2x"
                      width="280"
                    />
                  </picture>
                </div>
                <div
                  class="product-card__info"
                >
                  <div
                    class="rate product-card__rate"
                  >
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <p
                      class="visually-hidden"
                      data-testid="rating"
                    >
                      Рейтинг: 
                      5
                    </p>
                    <p
                      class="rate__count"
                      data-testid="rate-count"
                    >
                      <span
                        class="visually-hidden"
                      >
                        Всего оценок:
                      </span>
                      2
                    </p>
                  </div>
                  <p
                    class="product-card__title"
                    data-testid="name"
                  >
                    Van Shot
                  </p>
                  <p
                    class="product-card__price"
                    data-testid="price"
                  >
                    <span
                      class="visually-hidden"
                    >
                      Цена:
                    </span>
                    149 990
                     ₽
                  </p>
                </div>
                <div
                  class="product-card__buttons"
                >
                  <button
                    class="btn btn--purple product-card__btn"
                    type="button"
                  >
                    Купить
                  </button>
                  <a
                    class="btn btn--transparent"
                    href="/catalog/product/5"
                  >
                    Подробнее
                  </a>
                </div>
              </div>
              <div
                class="product-card is-active"
                data-testid="product-card"
              >
                <div
                  class="product-card__img"
                >
                  <picture>
                    <source
                      srcset="img/content/click-sap.webp, img/content/click-sap@2x.webp 2x"
                      type="image/webp"
                    />
                    <img
                      alt="Click Sap"
                      height="240"
                      src="img/content/click-sap.jpg"
                      srcset="img/content/click-sap@2x.jpg 2x"
                      width="280"
                    />
                  </picture>
                </div>
                <div
                  class="product-card__info"
                >
                  <div
                    class="rate product-card__rate"
                  >
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        data-testid="full-star-icon"
                        xlink:href="#icon-full-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        xlink:href="#icon-star"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      height="16"
                      width="17"
                    >
                      <use
                        xlink:href="#icon-star"
                      />
                    </svg>
                    <p
                      class="visually-hidden"
                      data-testid="rating"
                    >
                      Рейтинг: 
                      3
                    </p>
                    <p
                      class="rate__count"
                      data-testid="rate-count"
                    >
                      <span
                        class="visually-hidden"
                      >
                        Всего оценок:
                      </span>
                      3
                    </p>
                  </div>
                  <p
                    class="product-card__title"
                    data-testid="name"
                  >
                    Click Sap
                  </p>
                  <p
                    class="product-card__price"
                    data-testid="price"
                  >
                    <span
                      class="visually-hidden"
                    >
                      Цена:
                    </span>
                    9 490
                     ₽
                  </p>
                </div>
                <div
                  class="product-card__buttons"
                >
                  <button
                    class="btn btn--purple product-card__btn"
                    type="button"
                  >
                    Купить
                  </button>
                  <a
                    class="btn btn--transparent"
                    href="/catalog/product/6"
                  >
                    Подробнее
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
